specification.md

```diff
@@ -1518,6 +1602,438 @@ attributes or not. It is, therefore, entirely possible for
 
 </div>
 
+
+### StageInSet
+
+This class represents a set of stage-in directives, each represented by a 
+[`StageIn`](#stagein) object, that have no particular order. 
+
+<div class="imp-note">
+
+In languages with generic types and a `Set` type, a staging set could
+simply be `Set<StageIn>`, but implementations have the liberty to
```

Comment:
I would suggest to change 'implementations have the liberty' to 'language bindings have the liberty' to ensure that the same API is implemented by different implementations in the same language.

Reply:
Perhaps. I'm worried less about having to do `s/Set/StageInSet/` in select places than having to make algorithmic changes when switching from one implementation to another. Such a switch will most likely involve some changes. As long as the changes are simple and easily visible because an IDE underlines the wrong type, it is likely not a significant burden.

Reply:
I am more in the strict camp here.  Understand your point, and I assume in practice this is how it will go.  But it itches me that an implementor could come along and call her thing 'PSI/J compliant implementation' but I would need to figure out the subtle differences in the implementations.  Also, this seems to be the only point in the spec where we allow the implementor to choose a type - why here specifically?

Reply:
My comment should not block a merge - I will not veto based on this issue.

Reply:
> But it itches me that an implementor could come along and call her thing 'PSI/J compliant implementation' but I would need to figure out the subtle differences in the implementations. 

I don't quite see having to change `StageInSet` to `Set<StageIn>` everywhere as a particularly problematic thing. It's well defined as opposed to semantic changes, which I think are more subtle. So I think it's more important for the semantics to be constrained rather than the various syntactical/morphological aspects.

> Also, this seems to be the only point in the spec where we allow the implementor to choose a type - why here specifically?

We don't really have collections in many other places. There are the spec argument list and env dict. But here we have two sets with distinct member types.

I think the point here is that in pre-generics Java, I would have added a new class. In post-generics Java, not so much, as `Set<>` is good and simple enough. I'm not as sure in Python, but I'd also lean towards `Set[]`. But then Python is special, and you could always say `StageInSet = Set[StageIn]`. So it seems to me that some degree of flexibility is necessary here.

Reply:
We may be talking about different things, sorry if I wasn't clear.  Or maybe we are, but then I am confused.  Let me try to rephrase please.

My concern is application portability.  If Java implementation 1 provides the method "`StageInSet()`" and java implementation 2 does not, then an application needs code changes to switch from impl 1 to impl 2.  Those changes might be trivial - but I want to argue that they should not be needed in the first place.  I would thus suggest that the existence of `StageInSet()` is defined by the *language binding*, not by the *implementation*.

I have no opinion whatsoever on the specifics of expressing stagein set construction in Java.

Reply:
> We may be talking about different things, sorry if I wasn't clear. Or maybe we are, but then I am confused. Let me try to rephrase please.

Based on your answer, I think we are probably talking about the same thing.

> Those changes might be trivial - but I want to argue that they should not be needed in the first place. I would thus suggest that the existence of `StageInSet()` is defined by the _language binding_, not by the _implementation_.

I think my reasoning was that the changes are trivial and they are a lower cost than the combined cost of (possibly) overspecifying and needing language bindings for this. We can specify a binding for Java, but our bigger difficulty is specifying a binding for languages we don't know or that may not yet exist, since, with a Java language binding, we may solve this problem for Java, but for all other languages that might have a generic set, we will not.

So I do acknowledge the portability concern and it is indeed a concern. My view there is that switching PSI/J implementations would be a rare event in the lifecycle of a project AND that the cost would be low compared to the overall cost of making such a switch. In this particular case, I imagine that the change from `StageInSet` to `Set<StageIn>` would be a search/replace. More costly would probably be the cognitive load of determining that you need to do a search/replace for those types in the first place. However, on the latter point, I think such things would be inevitable, and any switch will involve some amount of testing and poking around.

Reply:
I've updated the wording a bit to mandate the use of a generic set if it is available and encourage what is typical of the language used when implementing a custom class. I don't think it feels fully right, but I do think it better addresses your concern.

I would probably move ahead with this and let feedback come from actual implementations.

Reply:
Thanks for spending the time to carefully think about this stuff and for a different perspective.